tracing/trace_state.hh

```diff
@@ -505,28 +505,118 @@ private:
 };
 
-class trace_state_ptr final {
+
+class opentelemetry_state final {
 private:
     lw_shared_ptr<trace_state> _state_ptr;
+    bool const _opentelemetry_tracing{false};
+
+public:
+    opentelemetry_state() = default;
+    opentelemetry_state(lw_shared_ptr<trace_state> state_ptr, bool opentelemetry_tracing = false)
+            : _state_ptr(std::move(state_ptr)), _opentelemetry_tracing(opentelemetry_tracing)
+    {}
+    opentelemetry_state(std::nullptr_t, bool opentelemetry_tracing = false)
+            : _state_ptr(nullptr), _opentelemetry_tracing(opentelemetry_tracing)
+    {}
+
+    bytes serialize() const noexcept {
+        bytes serialized{};
+        return serialized;
+    }
+
+    /**
+     * @return True if OpenTelemetry trace state is stored.
+     */
+    bool has_opentelemetry() const noexcept {
+        return _opentelemetry_tracing;
+    };
+
+    /**
+     * @return True if classic trace state is stored.
+     */
+    bool has_tracing() const noexcept {
+        return bool(_state_ptr);
+    };
+
+    /**
+     * @return A pointer to classic trace state.
+     */
+    trace_state* get_tracing_ptr() const noexcept {
+        return _state_ptr.get();
+    }
+
+    /**
+     * @return A reference to classic trace state.
+     */
+    trace_state& get_tracing() const noexcept {
+        return *_state_ptr;
+    }
+};
+
+
+class trace_state_ptr final {
+private:
+    lw_shared_ptr<opentelemetry_state> _state_ptr;
+
+    /**
+     * @return True if classic or OpenTelemetry trace state is stored.
+     */
+    bool has_any_tracing() const noexcept {
+        return __builtin_expect(bool(_state_ptr), false);
+    }
 
 public:
     trace_state_ptr() = default;
-    trace_state_ptr(lw_shared_ptr<trace_state> state_ptr)
+    trace_state_ptr(lw_shared_ptr<opentelemetry_state> state_ptr)
         : _state_ptr(std::move(state_ptr))
     {}
+    trace_state_ptr(lw_shared_ptr<trace_state> state_ptr)
+        : _state_ptr(make_lw_shared<opentelemetry_state>(std::move(state_ptr)))
+    {}
     trace_state_ptr(std::nullptr_t)
         : _state_ptr(nullptr)
     {}
 
-    explicit operator bool() const noexcept {
-        return __builtin_expect(bool(_state_ptr), false);
+    /**
+     * @return True if classic trace state is stored.
+     */
+    bool has_tracing() const noexcept {
+        return has_any_tracing() && _state_ptr->has_tracing();
+    };
+
+    /**
+     * @return A pointer to classic trace state.
+     */
+    trace_state* get_tracing_ptr() const noexcept {
+        return _state_ptr->get_tracing_ptr();
+    }
+
+    /**
+     * @return A reference to classic trace state.
+     */
+    trace_state& get_tracing() const noexcept {
+        return _state_ptr->get_tracing();
     }
 
-    trace_state* operator->() const noexcept {
+    /**
+     * @return True if OpenTelemetry trace state is stored.
+     */
+    bool has_opentelemetry() const noexcept {
+        return has_any_tracing() && _state_ptr->has_opentelemetry();
```

Comment:
```suggestion
        return __builtin_expect(has_any_tracing() && _state_ptr->has_opentelemetry(), false);
```

although the end goal is to have OpenTelemetry turned on all the time, this (should?) minimize the overhead when it's not turned on.